janusgraph-core/src/main/java/org/janusgraph/graphdb/query/index/IndexSelectionStrategy.java

```diff
@@ -0,0 +1,231 @@
+// Copyright 2020 JanusGraph Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package org.janusgraph.graphdb.query.index;
+
+import com.google.common.base.Preconditions;
+import org.janusgraph.core.JanusGraphElement;
+import org.janusgraph.core.JanusGraphRelation;
+import org.janusgraph.core.PropertyKey;
+import org.janusgraph.core.RelationType;
+import org.janusgraph.core.attribute.Cmp;
+import org.janusgraph.core.schema.JanusGraphSchemaType;
+import org.janusgraph.core.schema.SchemaStatus;
+import org.janusgraph.graphdb.database.IndexSerializer;
+import org.janusgraph.graphdb.internal.OrderList;
+import org.janusgraph.graphdb.query.QueryUtil;
+import org.janusgraph.graphdb.query.condition.*;
+import org.janusgraph.graphdb.query.graph.JointIndexQuery;
+import org.janusgraph.graphdb.types.*;
+import org.janusgraph.graphdb.types.system.ImplicitKey;
+import org.javatuples.Pair;
+
+import javax.annotation.Nullable;
+import java.util.*;
+import java.util.stream.StreamSupport;
+
+/**
+ * @author Florian Grieskamp (Florian.Grieskamp@gdata.de)
+ */
+public interface IndexSelectionStrategy {
+    Pair<JointIndexQuery, Boolean> selectIndices(final Set<IndexType> indexCandidates, final MultiCondition<JanusGraphElement> conditions,
```

Comment:
Instead of using `Pair` as a return type, I would create some POJO which contains `JointIndexQuery jointQuery` and `Boolean sorted` because 1) It isn't obvious what does `Boolean` means here. 2) If we decide to extend or shrink our return type, we won't be able to use Pair anymore.